torch_xla/csrc/xla_lower_util.cpp

```diff
@@ -1264,4 +1264,247 @@ xla::XlaOp BuildTpuCustomCall(const std::vector<xla::XlaOp>& inputs,
+        xla::XlaOp box_index = values[0];
+        // Check: current loop counter is within bounds, i.e. has a
+        // corresponding box.
+        return xla::Lt(box_index,
+                       xla::ConstantR0<IndexType>(builder, num_boxes));
+      },
+      [=](absl::Span<const xla::XlaOp> values, xla::XlaBuilder* builder) {
+        const xla::XlaOp ONE = xla::One(builder, XLAIndexType);
+        const xla::XlaOp ZERO = xla::Zero(builder, XLAIndexType);
+
+        xla::XlaOp box_index = values[0];
+        xla::XlaOp state = values[1];
+        xla::XlaOp iou_threshold_mask = values[2];
+
+        // Retrieve the IoU mask row corresponding to this box.
+        xla::XlaOp box_iou_threshold_mask = xla::DynamicSlice(
+            iou_threshold_mask, {box_index, ZERO}, {1, num_boxes});
+
+        // Update the current state with the IoU mask.
+        // Basically, sets to false every box X whose IoU with the current box
+        // is less-than or equal than the given threshold.
+        xla::XlaOp updated_state = xla::And(
+            state,
+            // Update the mask so that if we select this box
+            // (i.e. state[box] == true), we don't de-select it.
+            xla::DynamicUpdateSlice(
+                // Before that, we need to pre-process the mask.
+                //   1. Negate the mask: if this box is selected, we only want
+                //      those that have a low intersection ratio.
+                //   2. Reshape it to: [num_boxes].
+                xla::Reshape(xla::Not(box_iou_threshold_mask), {num_boxes}),
+                xla::ConstantR1<bool>(builder, {true}), {box_index}));
+
+        // Flag: should this box (loop counter) be included in the output?
+        xla::XlaOp should_include = xla::DynamicSlice(state, {box_index}, {1});
+        // Pick the new values of state, depending on whether we should include
+        // this box or not.
+        xla::XlaOp new_state =
+            xla::Select(xla::BroadcastInDim(should_include, {num_boxes}, {0}),
+                        updated_state, state);
+
+        xla::XlaOp next_box_index = box_index + ONE;
+        return std::vector<xla::XlaOp>{next_box_index, new_state,
+                                       iou_threshold_mask};
+      },
+      init_values, "BoxSelectionLoop", builder));
+}
+
+xla::XlaOp BuildNms(xla::XlaOp boxes, xla::XlaOp scores,
+                    xla::XlaOp iou_threshold) {
+  using IndexType = int32_t;
+  const xla::PrimitiveType XLAIndexType = xla::PrimitiveType::S32;
+
+  xla::XlaBuilder* builder = boxes.builder();
+
+  const int64_t COORDINATES = 4;
+  const xla::XlaOp ONE = xla::One(builder, XLAIndexType);
+  const xla::XlaOp ZERO = xla::Zero(builder, XLAIndexType);
+
+  const xla::Shape& boxes_shape = ShapeHelper::ShapeOfXlaOp(boxes);
+  XLA_CHECK_EQ(boxes_shape.rank(), 2);
+  XLA_CHECK_EQ(boxes_shape.dimensions(1), COORDINATES);
+  int64_t num_boxes = boxes_shape.dimensions(0);
```

Comment:
I wonder if the implementation here is a copy with some modification of the deleted nms_op.cpp, or if it is a brand new implementation.

Reply:
I would say that it is a new implementation heavily based on the previous one. So, the problems that I saw with the previous implementation were:

- Almost no comments: couldn't easily tell what some parts of the code were doing
- Copied from another source: it was copied from an old version of tensorflow
- Different signature and semantics: it returned the best `output_size` box indices, even though some of them might have been suppressed

I believe this implementation checks all of the above-mentioned boxes. I added plenty of comments + a few changes that resulted in a more maintainable code (IMHO).